src/controllers/csi/provisioner/controller.go

```diff
@@ -76,6 +83,34 @@ func (provisioner *OneAgentProvisioner) SetupWithManager(mgr ctrl.Manager) error
 }
 
 func (provisioner *OneAgentProvisioner) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	fs := afero.NewOsFs()
+	if provisioner.parallelDownloadsEnabled() {
+		if provisioner.parallelDownloadsLimitReached() {
+			log.Info("max parallel downloads reached, requeued")
+			return reconcile.Result{RequeueAfter: shortRequeueDuration}, nil
+		}
+		atomic.AddInt64(&provisioner.currentParallelDownloads, 1)
+		log.Info("staring parallel download")
+		go func() {
+			defer atomic.AddInt64(&provisioner.currentParallelDownloads, -1)
+			_, err := provisioner.reconcile(ctx, request, fs)
+			if err != nil {
+				log.Error(err, "Problem while provisioning oneagents in parallel")
+			}
+		}()
+		return reconcile.Result{RequeueAfter: defaultRequeueDuration}, nil
+	}
+	return provisioner.reconcile(ctx, request, fs)
+}
+
+func (provisioner *OneAgentProvisioner) parallelDownloadsEnabled() bool {
+	return provisioner.maxParallelDownloads > 0
+}
+func (provisioner *OneAgentProvisioner) parallelDownloadsLimitReached() bool {
+	return provisioner.currentParallelDownloads >= provisioner.maxParallelDownloads
+}
+
+func (provisioner *OneAgentProvisioner) reconcile(ctx context.Context, request reconcile.Request, fs afero.Fs) (reconcile.Result, error) {
```

Comment:
Why is the filesystem not part of the provisioner anymore?

Reply:
`afero` is not thread safe, (I don't know how or for which functions)
so to be safe each goroutine gets its own `fs` object